import-scripts/anonymize_age_at_seq_with_cap_py3.py

```diff
@@ -0,0 +1,421 @@
+        self.input_file_path = input_file_path
+        self.output_file_path = output_file_path
+        self.upper_age_limit = upper_age_limit
+        self.lower_age_limit = lower_age_limit
+        self.os_months_precision = os_months_precision
+        self.data_handler = DataHandler(input_file_path)
+        self.col_indices = self.data_handler.get_col_indices({'PATIENT_ID', 'OS_MONTHS', 'AGE_CURRENT'})
+        self.patient_os_years_map = {}
+
+    def write(self):
+        """Reads the patient clincal file, constructs a map from patient_id to os_years, applies age limits to AGE_CURRENT,
+        and writes the clinical patient file with os_months_precision"""
+        with open(self.input_file_path, 'r') as input_file_handle:
+            with open(self.output_file_path, 'w') as output_file_handle:
+                line_reader = PatientLineProcessor(
+                    self.patient_os_years_map,
+                    self.col_indices,
+                    output_file_handle,
+                    self.upper_age_limit,
+                    self.lower_age_limit,
+                    self.os_months_precision,
+                )
+                for line in input_file_handle:
+                    line_reader.process(line)
+
+
+class SampleLineProcessor(LineProcessor):
+    """Handles the processing of each line - capping ages according to limits"""
+
+    def __init__(self, patient_os_years_map, col_indices, output_file_handle, upper_age_limit, lower_age_limit):
+        super().__init__(patient_os_years_map, col_indices, output_file_handle)
+        self.upper_age_limit = upper_age_limit
+        self.lower_age_limit = lower_age_limit
+
+    def apply_age_column_limits(self, cols, age_col_index, patient_id_col_index):
+        """Applies age limits to the 'AGE_AT_SEQ_REPORTED_YEARS' column:
+            If 'AGE_AT_SEQ_REPORTED_YEARS' + OS_MONTHS (in years) is > upper_age_limit, this function will
+                overwrite the value of 'AGE_AT_SEQ_REPORTED_YEARS' to upper_age_limit - OS_MONTHS (in years).
+            If 'AGE_AT_SEQ_REPORTED_YEARS' is less than lower_age_limit, this function will throw a ValueError.
+
+        Args:
+            cols (list): Data values from the given line of the clinical sample file
+            age_col_index (int): Index of the 'AGE_AT_SEQ_REPORTED_YEARS' column
+            patient_id_col_index (int): Index of the 'PATIENT_ID' column
+
+        Raises:
+            ValueError: When the value at 'AGE_AT_SEQ_REPORTED_YEARS' is less than the lower age limit
+            IndexError: When a patient in the clinical sample file does not have a corresponding entry in the clinical patient file
+        """
+        age_value = cols[age_col_index]
+        patient_id_value = cols[patient_id_col_index]
+
+        # Don't process blank age values
+        if not age_value:
+            return
+
+        # Handle age values that are already capped
+        if age_value[0] == '>':
```

Comment:
I think this also will behave appropriately below. So with 89 as an upper age limit, if we had ">80" as our current value and had os_years of 3 we would simply leave the value of ">80" and not change it. However, if os_years became 10 the sum will be 90 so we would be replacing the old ">80" value with (89 - 10) or ">79". When os_years became 20 the sum will be 100 so we would be replacing the old ">80" value with 89 - 20 or ">69"

Reply:
@sheridancbio Yes, this sounds correct to me! I have some unit tests that test these cases as well.